**Task 1.**
Spend some time reviewing the datasets in the above link. Pick 3 of them, that look interesting to you. One of them must be relatively small (at most 5K nodes), and one of them must be a directed graph. Try to understand their format: these are text files, but how are these text files organized in order to signify nodes and edges?

Task #1 Answer:

Facebook-
The Facebook dataset contains nodes and edges that are combined in the, facebook_combined.txt, file. 

Github-
The Github dataset contains edges that are stored in, musae_git_edges.csv, and nodes that are stored in, musae_git_target.csv.

Twitter-
The Twitter dataset contains nodes and edges that are combined in the, twitter_combined.txt, file. 

**Task 2.**
Find or write a small interface that you can use to import these data sets into your language environment. The graph must been coded as a list of edges, where each edge is a pair of nodes.

In [1]:
import csv
import networkx as nx
import matplotlib.pyplot as plt
import tarfile
import pandas as pd
import sys 
import numpy as np
import operator
from collections import Counter

In [2]:
# Facebook
from google.colab import files
uploaded = files.upload()

Saving facebook_combined.txt to facebook_combined.txt


In [3]:
# GitHub - Edges
from google.colab import files
uploaded = files.upload()

Saving musae_git_edges.csv to musae_git_edges.csv


In [4]:
# GitHub - Nodes
from google.colab import files
uploaded = files.upload()

Saving musae_git_target.csv to musae_git_target.csv


In [5]:
# Twitter
from google.colab import files
uploaded = files.upload()

Saving twitter_combined.txt to twitter_combined.txt


In [7]:
G_fb = nx.read_edgelist('facebook_combined.txt')
print(nx.info(G_fb))

Name: 
Type: Graph
Number of nodes: 4039
Number of edges: 88234
Average degree:  43.6910


In [8]:
G_tw = nx.read_edgelist('twitter_combined.txt')
print(nx.info(G_tw))

Name: 
Type: Graph
Number of nodes: 81306
Number of edges: 1342310
Average degree:  33.0187


In [9]:
#G_gh = nx.read_edgelist('musae_git_edges.csv')

# Read in the nodelist file
with open('musae_git_target.csv', 'r') as nodecsv:
    nodereader = csv.reader(nodecsv)
    nodes = [n for n in nodereader][1:]

# Get a list of just the node names (the first item in each row)
node_names = [n[0] for n in nodes]

# Read in the edgelist file
with open('musae_git_edges.csv', 'r') as edgecsv:
    edgereader = csv.reader(edgecsv)
    edges = [tuple(e) for e in edgereader][1:]

G_gh = nx.Graph()                # Initialize a Graph object
G_gh.add_nodes_from(node_names)  # Add nodes to the Graph
G_gh.add_edges_from(edges)       # Add edges to the Graph
print(nx.info(G_gh))

Name: 
Type: Graph
Number of nodes: 37700
Number of edges: 289003
Average degree:  15.3317


**Milestone 1**
Start your writeup with a discussion of the characteristics of the data sets you picked, i.e. size, where they come from etc. (5 days) 

The datasets I picked to examine for this Mini Project are the Social circles from Facebook (anonymized), the Social circles from Twitter, and the Social network of Github developers. 

Facebook - 
1. Characteristics - The Facebook dataset is Undirected. This dataset consists of 'circles' (or 'friends lists') from Facebook. There are 10 ego-networks, consisting of 4,039 users.
2. Size - 88234 Edges, and 4039 Nodes. The size of the file, facebook_combined.txt, is 835 KB. 
3. Where they come from - Data was collected from survey participants using the Facebook app.

Twitter - 
1. Characteristics - The Twitter dataset is Directed. 
2. Size - 1768149 Edges, and 81306 Nodes. The size of the file, facebook_combined.txt, is 43,506 KB. 
3. Where they come from - Data was collected from being crawled from public sources. 

GitHub - 
1. Characteristics - The GitHub dataset is Undirected. The Nodes of the dataset are developers who have starred at least 10 repositories, and the edges are mutual follower relationships between them. The vertex features are extracted based on the location, repositories starred, employer, and e-mail address.
2. Size - 289003 Edges, and 37700 Nodes. The size of the file, musae_git_edges.csv, is 3229 KB, and the size of file, musae_git_target.csv, is 661 KB. 
3. Where they come from - Data was collected from a public API in June 2019. 

**2. Finding Influencers. Task 3.**
One interesting aspect of social networks is the existence of ‘influencers’. In a graph these are the nodes with high in-degree. For this task you will have to write a function that finds the 100 of the nodes with highest in-degree and report them, along with the number of their followers.

Degree - This represents the number of connections a node has. 

Degree centrality - This is a measure of the number of connections a particular node has in the network.

Eigenvector centrality - This decides if a node is important if it is connected to other important nodes.

Betweenness Centrality - This is the centrality of control. It represents the frequency at which a point occurs on the geodesic (shortest paths) that connected pair of points. 

In [ ]:
#G_fb.degree()

In [ ]:
#list(G_fb.nodes())
#dict(G_fb.degree())

In [ ]:
#G_fb.nodes(data=True)
#Counter(nx.eigenvector_centrality(G)).most_common(10)

In [10]:
nx.set_node_attributes(G_fb, dict(G_fb.degree()), 'Degree')
nx.set_node_attributes(G_fb, nx.degree_centrality(G_fb), 'Degree Centrality')
nx.set_node_attributes(G_fb, nx.eigenvector_centrality(G_fb), 'Eigenvector Centrality')
nx.set_node_attributes(G_fb, nx.betweenness_centrality(G_fb), 'Betweenness Centrality')

In [11]:
pd.DataFrame.from_dict(
 dict(G_fb.nodes(data=True)),
 orient='index'
).sort_values('Betweenness Centrality', ascending=False).head(100)

Degree  Degree Centrality  Eigenvector Centrality  Betweenness Centrality
107     1045           0.258791            2.606940e-04                0.480518
1684     792           0.196137            7.164260e-06                0.337797
3437     547           0.135463            9.531613e-08                0.236115
1912     755           0.186974            9.540696e-02                0.229295
1085      66           0.016345            3.164082e-06                0.149015
...      ...                ...                     ...                     ...
2189     113           0.027984            1.198784e-02                0.001003
606       91           0.022536            1.073341e-05                0.000997
774       37           0.009163            8.612631e-12                0.000987
322       72           0.017831            2.689807e-05                0.000964
1173     115           0.028479            8.957207e-05                0.000942

[100 rows x 4 columns]

In [13]:
nx.set_node_attributes(G_tw, dict(G_tw.degree()), 'Degree')
#nx.set_node_attributes(G_tw, nx.degree_centrality(G_tw), 'Degree Centrality')
#nx.set_node_attributes(G_tw, nx.eigenvector_centrality(G_tw), 'Eigenvector Centrality')
nx.set_node_attributes(G_tw, nx.betweenness_centrality(G_tw), 'Betweenness Centrality')

KeyboardInterrupt: ignored

In [ ]:
pd.DataFrame.from_dict(
 dict(G_tw.nodes(data=True)),
 orient='index'
).sort_values('Betweenness Centrality', ascending=False).head(100)

In [ ]:
nx.set_node_attributes(G_gh, dict(G_gh.degree()), 'Degree')
#nx.set_node_attributes(G_gh, nx.degree_centrality(G_gh), 'Degree Centrality')
#nx.set_node_attributes(G_gh, nx.eigenvector_centrality(G_gh), 'Eigenvector Centrality')
nx.set_node_attributes(G_gh, nx.betweenness_centrality(G_gh), 'Betweenness Centrality')

In [ ]:
pd.DataFrame.from_dict(
 dict(G_gh.nodes(data=True)),
 orient='index'
).sort_values('Betweenness Centrality', ascending=False).head(100)

**3. Cleaning-up the network.**
In a social network people and bots keep opening fake accounts. Such accounts are often ‘isolated’: they are disconnected from the bulk of the network. In graph terms, they may be either non-connected nodes, or small groups of nodes that have some connections to each other but they are not connected to the ‘main’ part of the network. 

**Task 4.**
Write code that will find and report these smaller components of the network. More specifically, you should write code that reports the (simple) connected components of a graph. The output must be an array C with n entries, where n is the number of the nodes in the graph. Each connected component should have an id number. Then, C[i] must be equal to the ID number of the connected component that contains node i. Your code should also output the list of edges comprising the biggest connected component. Note: You should treat the graph as undirected in this part.

Task #4 Answer: 

Modularity is a measure of relative density in your network: a community (called a module or modularity class) has high density relative to other nodes within its module but low density with those outside.

In [16]:
from networkx.algorithms import community
communities = community.greedy_modularity_communities(G_fb)

In [17]:
modularity_dict = {}                                          # Create a blank dictionary
for i,c in enumerate(communities):                            # Loop through the list of communities, keeping track of the number for the community
    for name in c:                                            # Loop through each person in a community
        modularity_dict[name] = i                             # Create an entry in the dictionary for the person, where the value is which group they belong to.

nx.set_node_attributes(G_fb, modularity_dict, 'modularity')   # Add modularity information 

In [18]:
def communitiesFb(n):
 for i,c in enumerate(communities):          # Loop through the list of communities
    if len(c) > n:                           # Filter out modularity classes fewer than n nodes
        print('C_id= ' + str(i) +',','n=',len(c),list(c))    # Print out the classes and their members

In [19]:
communitiesFb(30)

C_id= 0, n= 983 ['1421', '2954', '3350', '3280', '3326', '1098', '1639', '1662', '1003', '1609', '1401', '1476', '3225', '1867', '3188', '2824', '3218', '1538', '1494', '1822', '2904', '3245', '3159', '3196', '3378', '3035', '3019', '3404', '3375', '1615', '1719', '1736', '2957', '1806', '2691', '3295', '1187', '995', '3338', '1909', '1239', '3093', '1302', '1678', '1471', '2950', '1911', '3165', '1879', '1799', '2775', '1050', '3133', '1344', '1852', '1110', '3285', '1746', '3386', '3162', '1803', '3369', '2826', '3124', '3075', '2856', '1172', '3146', '1132', '1289', '3168', '1078', '1641', '1278', '3331', '3239', '1160', '1524', '1735', '3057', '1610', '1908', '2990', '1553', '1708', '2675', '2682', '3214', '2822', '1351', '2772', '1150', '2845', '1116', '1866', '1774', '1832', '2744', '1222', '1900', '1250', '3039', '3292', '2770', '928', '3050', '3211', '3230', '3287', '1209', '1440', '3041', '2665', '3083', '3138', '1173', '1723', '1055', '1619', '1101', '3354', '1849', '1579', '

**Milestone 2.** Complete Tasks 3-4. (10 days.)

**4. Closed Subnetworks.** 
In a Twitter-like social network there are subsets of people forming ‘closed sub networks’. Such people do not follow anyone else outside their sub network, and only follow other people within their sub network (but not all of them). People outside that sub network do follow people that are part of the subnetwork. For example, famous people on Twitter tend to not follow too many other people, except other famous people. So,they may be forming sub networks of famous people. 

**Task 5.** 

Write code that finds and reports all closed subnetworks of a Twitter-like network. More specifically, write code that computes the decomposition of the network into strongly connected components. The output must be an array S with n entries, where n is the number of the nodes in the graph. Each strongly connected component should have an id number. Then, S[i] must be equal to the ID number of the strongly connected component that contains node i.

Using Kosaraju’s algorithm, we can find strongly connected components in in O(V+E) time. The algorithm calls DFS (Depth first search), finds the reverse of the graph and again calls DFS. DFS takes O(V+E) for a graph represented using the adjacency list. Reversing a graph also takes O(V+E) time. For reversing the graph, we simple traverse all adjacency lists.

In [25]:
from collections import defaultdict 
  
class Graph: 

    def __init__(self,vertices): 
        self.V= vertices                 # # of vertices         
        self.graph = defaultdict(list)   # default dictionary to store graph          
        self.Time = 0
   
    def addEdge(self,u,v):               # function to add an edge to graph 
        self.graph[u].append(v) 

    def SCCUtil(self,u, low, disc, stackMember, st): 
        disc[u] = self.Time             # Initialize discovery time and low value 
        low[u] = self.Time 
        self.Time += 1
        stackMember[u] = True
        st.append(u) 
  
        for v in self.graph[u]:               
            if disc[v] == -1 :               
                self.SCCUtil(v, low, disc, stackMember, st) 
                low[u] = min(low[u], low[v])                          
            elif stackMember[v] == True:   
                low[u] = min(low[u], disc[v]) 
  
        w = -1 # To store stack extracted vertices 
        gh=[]
        if low[u] == disc[u]: 
            while w != u: 
                w = st.pop() 
                gh.append(w)
                stackMember[w] = False 
            print(gh)                  
              
    def SCC(self): 
        disc = [-1] * (self.V) 
        low = [-1] * (self.V) 
        stackMember = [False] * (self.V) 
        st =[] 
          
        for i in range(self.V): 
            if disc[i] == -1: 
                self.SCCUtil(i, low, disc, stackMember, st)   

In [33]:
g1 = Graph(6) 
g1.addEdge(1, 2) 
g1.addEdge(0, 1) 
g1.addEdge(2, 1) 
g1.addEdge(1, 0) 
g1.addEdge(3, 4) 
g1.addEdge(4, 5)
print("The strongly connected components in the graph are: ")
g1.SCC() 

The strongly connected components in the graph are: 
[2, 1, 0]
[5]
[4]
[3]


**4. A recommendation engine.** 

Three persons P1, P2, P3 form a length-3 following chain if a person P1 follows person P2, and person P2 follows P3. Given a length-3 following chain, it may make sense to recommend to P1 to follow person P3. It may make even more sense doing that if there are several 3- following chains between P1 and P3. 

**Task 6.** 

Write code that takes as input two persons P and P0 and a number k, finds the number of length-k following chains that connect P to P0.

In the below code, we can accomplish this. The code finds the number of paths of length-k in a directed graph. N represents the vertices, and k represents the length of the path. The graph is represented by the adjacency matrix where G[i][j]=1 is an edge from vertex i to j, and G[i][j]=0 represents no edge. 

The time complexity is O((|V|^2)*log(k)), where V is the number of vertices and k is the length of the path. This is the time complexity since we multiply the adjacency matrix log(k) times. 

In [90]:
N = 2  # Vertices- Person P and P0 in this case. 
def multiply(a, b, res) :    
    mul = np.zeros((N,N));        
    for i in range(N) : 
        for j in range(N) : 
            mul[i][j] = 0;  
            for k in range(N) :  
                mul[i][j] += a[i][k] * b[k][j];  
    for i in range(N) : 
        for j in range(N) : 
            res[i][j] = mul[i][j];  
  
def power(G, res, n) :  
    if (n == 1) : 
        for i in range(N) :  
            for j in range(N) : 
                res[i][j] = G[i][j];  
        return;  
    power(G, res, n // 2);   
    multiply(G, G, res);  
    if (n % 2 != 0) : 
        multiply(res, G, res);  

In [91]:
if __name__ == "__main__" :  
    G = [[ 1, 1 ],  
         [ 0, 0 ]];   
    k = 2;  # Length k
    res = np.zeros((N,N));    
    power(G, res, k);   
    for i in range(N) : 
        for j in range(N) : 
            print(res[i][j],end = " ");           
        print() 

1.0 1.0 
0.0 0.0 


Number of paths from 0 to 0 of length k is 1({0->0->0})

Number of paths from 0 to 1 of length k are 1({0->0->1})

**Milestone 3.** Complete tasks 5-6. (10 days)

Grading Notes: Milestone 1 and 2 should be completed by all. We have already essentially done these tasks. For milestone 3, students in the alternative track can use Python package networkx. This should be relatively easier, as the only thing required is to read the manual of some ready-made functions, and apply them on your data. For other students, task 5 requires writing code for finding a DFS of a graph. If you want to avoid that, you can also use networkxx for a deduction of 2/20 points. But the rest of the implementation should be yours